File: backend/src/AnimeShop/API/Stock/Manufacturers.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Stock.Manufacturers (ManufacturersAPI, manufacturersServer) where

import AnimeShop.Prelude

type ManufacturersAPI =
  "manufacturers" :>
  (
    Get '[JSON] [Entity Manufacturer] :<|>
    ReqBody '[JSON] Manufacturer :> Post '[JSON] (Entity Manufacturer) :<|>
    ReqBody '[JSON] ManufacturerId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity Manufacturer) :> Patch '[JSON] ()
  )

manufacturersServer :: ServerT ManufacturersAPI App
manufacturersServer = getManufacturers :<|> postManufacturer :<|> deleteManufacturer :<|> patchManufacturer

getManufacturers :: App [Entity Manufacturer]
getManufacturers = runDb $ selectList [] []

postManufacturer :: Manufacturer -> App (Entity Manufacturer)
postManufacturer = runDb . insertEntity

deleteManufacturer :: ManufacturerId -> App ()
deleteManufacturer = runDb . delete

patchManufacturer :: Entity Manufacturer -> App ()
patchManufacturer entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/manufacturers.js
import { ADD_MANUFACTURER, GET_MANUFACTURERS, DELETE_MANUFACTURER } from '../actions/types';

export function manufacturers(state = [], action) {
  switch (action.type) {
    case ADD_MANUFACTURER:
      return [ ...state, action.manufacturer];
    case GET_MANUFACTURERS:
      return action.manufacturers;
    case DELETE_MANUFACTURER:
      return state.filter(manufacturer => manufacturer.id !== action.id);
    default:
      return state;
  }
}
